pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,971 @@
+		Self::take_transaction_fee_from_user(schedule_id, schedule, weight_for_dca_execution)?;
+
+		if Self::price_change_is_bigger_than_max_allowed(schedule) {
+			Self::retry_schedule(schedule_id, schedule, current_blocknumber)?;
+
+			return Err(Error::<T>::PriceChangeIsBiggerThanMaxAllowed.into());
+		}
+
+		Ok(())
+	}
+
+	#[transactional]
+	pub fn execute_trade(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+	) -> DispatchResult {
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+
+		match &schedule.order {
+			Order::Sell {
+				asset_in,
+				asset_out,
+				amount_in,
+				min_limit,
+				slippage,
+				route,
+			} => {
+				let remaining_amount_to_use =
+					RemainingAmounts::<T>::get(schedule_id).defensive_ok_or(Error::<T>::InvalidState)?;
+				let amount_to_sell = min(remaining_amount_to_use, *amount_in);
+
+				Self::unallocate_amount(schedule_id, schedule, amount_to_sell)?;
+
+				let (estimated_amount_out, slippage_amount) =
+					Self::calculate_estimated_and_slippage_amounts(*asset_out, *asset_in, amount_to_sell, *slippage)?;
+				let min_limit_with_slippage = estimated_amount_out
+					.checked_sub(slippage_amount)
+					.ok_or(ArithmeticError::Overflow)?;
+				let min_limit = max(*min_limit, min_limit_with_slippage);
+
+				let route = route.to_vec();
+				let trade_amounts =
+					pallet_route_executor::Pallet::<T>::calculate_sell_trade_amounts(&route, amount_to_sell.into())?;
+				let last_trade = trade_amounts.last().defensive_ok_or(Error::<T>::InvalidState)?;
+				let amount_out = last_trade.amount_out;
+
+				ensure!(amount_out >= min_limit.into(), Error::<T>::TradeLimitReached);
+
+				pallet_route_executor::Pallet::<T>::sell(
+					origin,
+					*asset_in,
+					*asset_out,
+					(amount_to_sell).into(),
+					min_limit.into(),
+					route,
+				)
+			}
+			Order::Buy {
+				asset_in,
+				asset_out,
+				amount_out,
+				slippage,
+				max_limit,
+				route,
+			} => {
+				let amount_in = Self::get_amount_in_for_buy(amount_out, route)?;
+
+				Self::unallocate_amount(schedule_id, schedule, amount_in)?;
+
+				let (estimated_amount_in, slippage_amount) =
+					Self::calculate_estimated_and_slippage_amounts(*asset_in, *asset_out, *amount_out, *slippage)?;
+				let max_limit_with_slippage = estimated_amount_in
+					.checked_add(slippage_amount)
+					.ok_or(ArithmeticError::Overflow)?;
+
+				let max_limit = min(*max_limit, max_limit_with_slippage);
+				ensure!(amount_in <= max_limit, Error::<T>::TradeLimitReached);
+
+				pallet_route_executor::Pallet::<T>::buy(
+					origin,
+					*asset_in,
+					*asset_out,
+					(*amount_out).into(),
+					max_limit.into(),
+					route.to_vec(),
+				)
+			}
+		}
+	}
+
+	fn replan_or_complete(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+		current_blocknumber: T::BlockNumber,
+	) -> DispatchResult {
+		Self::deposit_event(Event::TradeExecuted {
```

Comment:
exact amount in and amount out that user sent and received has be here for UI to be able to show it to users

Reply:
for this you will have to return these values from execute trade and move emit to the parent fn

Reply:
consider to return these as well in trade failed if available

Reply:
for the trade executed, it is done, on the other hand for the failing scenario, these values are not always available.

Even when they are available later on, it would complicate the error handling, so instead of returning `Result<(Balance, Balance), DispatchError>`, we would need to return something like `Result<(Balance, Balance), (DispatchError, (Balance, Balance))>` which would lead to a lot of error mapping and complexity.

The `execute_trade` method is also transactional, so it should return DispatchError as Error.

If a trade fails with an error, the dispatch error will be in the `TradeFailed` event.